contrib/http_base/src/main/java/io/opentelemetry/contrib/http/core/HttpClientHandler.java

```diff
@@ -0,0 +1,183 @@
+ * limitations under the License.
+ */
+
+package io.opentelemetry.contrib.http.core;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+import static io.opentelemetry.contrib.http.core.HttpTraceConstants.INSTRUMENTATION_LIB_ID;
+
+import io.opentelemetry.OpenTelemetry;
+import io.opentelemetry.context.propagation.HttpTextFormat;
+import io.opentelemetry.distributedcontext.DistributedContext;
+import io.opentelemetry.distributedcontext.DistributedContextManager;
+import io.opentelemetry.distributedcontext.Entry;
+import io.opentelemetry.distributedcontext.EntryMetadata.EntryTtl;
+import io.opentelemetry.metrics.Meter;
+import io.opentelemetry.trace.Span;
+import io.opentelemetry.trace.Span.Kind;
+import io.opentelemetry.trace.SpanContext;
+import io.opentelemetry.trace.Tracer;
+import java.util.logging.Logger;
+import javax.annotation.Nullable;
+
+/**
+ * This helper class provides routine methods to instrument HTTP servers.
+ *
+ * @param <Q> the HTTP request entity.
+ * @param <P> the HTTP response entity.
+ * @param <C> the type of the carrier.
+ */
+public class HttpClientHandler<Q, P, C> extends AbstractHttpHandler<Q, P> {
+
+  private static final Logger LOGGER = Logger.getLogger(HttpClientHandler.class.getName());
+
+  private final Tracer tracer;
+  private final DistributedContextManager contextManager;
+  private final HttpTextFormat.Setter<C> setter;
+  private final HttpTextFormat<SpanContext> spanContextHttpTextFormat;
+  private final HttpTextFormat<DistributedContext> distributedContextHttpTextFormat;
+
+  /**
+   * Constructs a handler object.
+   *
+   * @param extractor used to extract information from request/response
+   * @param setter the setter used to extract information from the carrier
+   */
+  public HttpClientHandler(HttpExtractor<Q, P> extractor, HttpTextFormat.Setter<C> setter) {
+    this(extractor, setter, null, null, null, null);
+  }
+
+  /**
+   * Constructs a handler object.
+   *
+   * @param extractor used to extract information from request/response
+   * @param setter the setter used to extract information from the carrier
+   * @param statusConverter the HTTP status to OT status translator
+   * @param tracer the OT tracer
+   * @param contextManager the OT distributed context manager
+   * @param meter the OT meter
+   */
+  public HttpClientHandler(
+      HttpExtractor<Q, P> extractor,
+      HttpTextFormat.Setter<C> setter,
+      StatusCodeConverter statusConverter,
+      Tracer tracer,
+      DistributedContextManager contextManager,
+      Meter meter) {
+    super(extractor, statusConverter, meter);
+    checkNotNull(setter, "setter is required");
+    if (tracer == null) {
+      this.tracer = OpenTelemetry.getTracerFactory().get(INSTRUMENTATION_LIB_ID);
+    } else {
+      this.tracer = tracer;
+    }
+    if (contextManager == null) {
+      this.contextManager = OpenTelemetry.getDistributedContextManager();
+    } else {
+      this.contextManager = contextManager;
+    }
```

Comment:
these if/elses should be handled via different constructors, to make the intent clear.

Reply:
Improved constructors and documentation

Reply:
you shouldn't need the if/elses if the constructors are sorted out.

Reply:
Changed